### PR TITLE
#15898 change ActorPath to abstract class to make the methods in the com...

### PR DESCRIPTION
**Pull request by He-Pin**
1,make the ValidSymbols private
2,change the ActorPath from sealed trait to sealed abstract class
3,and an overload method for validatePathElement

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 1]
```diff
@@ -28,9 +28,17 @@ object ActorPath {
...
+
+  /**
+   * Validates the given actor path element and throws an [[InvalidActorNameException]] if invalid.
+   * See [[isValidPathElement()]] for a non-throwing version.
+   *
+   * @param element actor path element to be validated
    * @param fullPath optional fullPath element that may be included for better error messages; null if not given
    */
-  final def validatePathElement(element: String, fullPath: String = null): Unit = {
+  final def validatePathElement(element: String, fullPath: String): Unit = {
```
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: This change is not binary compatible (unfortunate, I know).

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 2]
```diff
@@ -98,7 +106,7 @@ object ActorPath {
  * when comparing actor paths.
  */
 @SerialVersionUID(1L)
-sealed trait ActorPath extends Comparable[ActorPath] with Serializable {
+sealed abstract class ActorPath extends Comparable[ActorPath] with Serializable {
```
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: we’ll see what MiMa says to this one, might work

**Comment by rkuhn** [comment]
OK TO TEST

**Comment by rkuhn** [comment]
refs #15898 

Thanks for the PR! Concerning the BC compatibility comment: it should be fine to add that overload for Java and make it `protected` (unless that removes the static forwarder, please try it out).

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 3]
```diff
@@ -18,7 +18,7 @@ object ActorPath {
   }
 
   /** INTERNAL API */
-  private[akka] final val ValidSymbols = """-_.*$+:@&=,!~';"""
+  private final val ValidSymbols = """-_.*$+:@&=,!~';"""
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: if `private` the `/** INTERNAL API */` marker can be removed

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala (hunk 1 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: IMO good news - this is a new method on master (I have added it in https://github.com/akka/akka/commit/c31469dd2930669324512a7750ecc1cb8e414d8f , on 2.3 we only have `isValidPathElement`) and it's on an object - not a trait, so we're safe from the "method added to trait" problem.

I think we might get away with this change in this case.

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala (hunk 2 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: We don't have mima enabled on `master`. 
I think we're currently not passing MiMa between 2.3 and 2.4 - @patriknw did experiment towards this but we didn't resolve all incompatibilities AFAIR

Comment (rkuhn) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: Yes, I know; I’m just keeping a watchful eye on these as we will have to correct all the mistakes in a few weeks time anyway.

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala (hunk 1 above)
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala]: ah, then all is good, thanks (and sorry for not checking)

**Comment by rkuhn** [comment]
LGTM, thanks!

**Pull request merged**